aiida_vasp/calcs/workchains/converge.py

```diff
@@ -0,0 +1,1392 @@
+            if settings['k_samples'] is None:
+                if self._verbose:
+                    self.report('setting the default value of the number of k-point samples '
+                                'to perform in the convergence test (k_samples) to {}'.format(self._k_samples_default))
+                settings['k_samples'] = self._k_samples_default
+            if settings['k_spacing'] is None:
+                if self._verbose:
+                    self.report('setting the default value of the k-point spacing ' '(k_spacing) to {}'.format(self._k_spacing_default))
+                settings['k_spacing'] = self._k_spacing_default
+
+        return
+
+    def _check_context_cutoffs(self):
+        """Check and verify cutoff related parameters."""
+        settings = self.ctx.converge.settings
+
+        if settings['cutoff_type'] is None:
+            if self._verbose:
+                self.report('setting the default convergence parameter to the default ' '({})'.format(self._cutoff_type_default))
+            settings['cutoff_type'] = self._cutoff_type_default
+        try:
+            self._ALLOWED_CUTOFF_TYPES[settings['cutoff_type']]
+        except KeyError:
+            if self._verbose:
+                self.report('the supplied convergence parameter is not supported, '
+                            'setting it to the default ({})'.format(self._cutoff_type_default))
+            settings['cutoff_type'] = self._cutoff_type_default
+        if settings['cutoff_value'] is None:
+            if self._verbose:
+                self.report('setting the default convergence value to {}'.format(self._cutoff_value_default))
+            settings['cutoff_value'] = self._cutoff_value_default
+        if settings['cutoff_value_r'] is None:
+            if self._verbose:
+                self.report('setting the default relative convergence value to {}'.format(self._cutoff_value_r_default))
+            settings['cutoff_value_r'] = self._cutoff_value_r_default
+
+        return
+
+    def _check_context_relative(self):
+        """Check and verify existence of relative convergence tests."""
+        settings = self.ctx.converge.settings
+
+        if settings['compress'] is None:
+            if self._verbose:
+                self.report('turning the relative convergence test for compression on')
+            settings['compress'] = True
+        if settings['displace'] is None:
+            if self._verbose:
+                self.report('turning the relative convergence test for displacement on')
+            settings['displace'] = True
+        self._check_context_relative_st()
+
+        return
+
+    def _check_context_relative_st(self):
+        """Check and verify parameters related to the relative convergence tests."""
+        settings = self.ctx.converge.settings
+
+        # Only care about displacement parameters if relative displacement convergence tests are performed.
+        if settings['displace']:
```

Comment:
It looks like a lot of repetition with only small changes to me, would it make sense to write a function that checks, sets the default if necessary and reports in one go?, then potentially you could loop over `settings` and make this and related functions much shorter.

Reply:
Yes, I can put that on the issue list. I thought about cleaning this up.